poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+	p.pubKey     = keyPair.Public
+	p.guardians  = guardians
+	p.secrets    = make([]abstract.Secret, p.n , p.n )
+	p.signatures = make([]*PromiseSignature, p.n , p.n )
+
+	// Verify that t <= r <= n
+	if p.n  < p.t {
+		panic("Not enough guardians for the secret")
+	} 
+	if p.r < p.t {
+		p.r = p.t
+	}
+	if p.r > p.n {
+		p.r = p.n
+	}
+
+	// Create the public polynomial and private shares. The total shares made
+	// should be equal to teh number of guardians while the minimum shares
+	// needed to reconstruct should be t.
+	pripoly   := new(poly.PriPoly).Pick(p.shareSuite, p.t, keyPair.Secret, random.Stream)
+	prishares := new(poly.PriShares).Split(pripoly, p.n)
+	p.pubPoly = new(poly.PubPoly).Commit(pripoly, nil)
+	
+	// Create an empty PriShares for the promised.
+	p.numShares = 0
+	p.priShares = new(poly.PriShares)
+	p.priShares.Empty(p.shareSuite, p.t, p.n)
+	
+	// Populate the secrets array. It encrypts each share with a diffie
+	// hellman exchange between the originator of the promist and the
+	// specific guardian.
+	for i := 0 ; i < p.n; i++ {
+		diffieBase := p.shareSuite.Point().Mul(guardians[i], keyPair.Secret)
+		p.secrets[i] = p.diffieHellmanEncrypt(prishares.Share(i), diffieBase)
+	}
+	
+	return p
+}
+
+// Returns the id of the policy
+func (p *Promise) GetId() string {
+	return p.id
+}
+
+/* Given a Diffie-Hellman shared key, encrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to encrypt
+ *    diffieBase  = the DH shared key
+ *
+ * Return
+ *   the encrypted secret
+ */
+func (p *Promise) diffieHellmanEncrypt(secret abstract.Secret, diffieBase abstract.Point) abstract.Secret {	
+	buff, err := diffieBase.MarshalBinary()
+	if err != nil {
+		panic("Bad shared secret for Diffie-Hellman give.")
+	}
+	cipher := p.shareSuite.Cipher(buff)
+	diffieSecret := p.shareSuite.Secret().Pick(cipher)
+	return p.shareSuite.Secret().Add(secret, diffieSecret)
+}
+
+/* Given a Diffie-Hellman shared key, decrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to decrypt
+ *    diffieBase  = the DH shared key
+ *
+ * Return
+ *   the decrypted secret
+ */
+func (p *Promise) diffieHellmanDecrypt(secret abstract.Secret, diffieBase abstract.Point) abstract.Secret {	
+	buff, err := diffieBase.MarshalBinary()
+	if err != nil {
+		panic("Bad shared secret for Diffie-Hellman give.")
+	}
+	cipher := p.shareSuite.Cipher(buff)
+	diffieSecret := p.shareSuite.Secret().Pick(cipher)
+	return p.shareSuite.Secret().Sub(secret, diffieSecret)
+}
+
+/* This helper function makes sure that a share is "syntactically" valid. In
+ * other words, the index is properly in range and the guardian it was sent to
+ * the correct guardian. VerifyShare handles "semantic" verification.
+ *
+ * Arguments
+ *   i        = the index of the share
+ *   gKeyPair = the key pair of the guardian of the share
+ *
+ * Returns
+ *    Whether the share is valid or invalid.
+ */
+func (p *Promise) validShare(i int, gKeyPair *config.KeyPair) bool {
```

Comment:
Where is the actual validation of the share here?  I don't even see reference to an actual "share" in this function or its arguments, only a share index.